freebsd/hid.c

```diff
@@ -0,0 +1,1143 @@
+ * The algorithm should give similar results as Apple's:
+ *   https://developer.apple.com/documentation/iokit/kiohiddeviceusagepairskey?language=objc
+ * Physical Collections are also matched (macOS does the same).
+ *
+ * This function can be called repeatedly until it returns non-0
+ * Usage is found. pos is the starting point (initially 0) and will be updated
+ * to the next search position.
+ *
+ * The return value is 0 when a pair is found.
+ * 1 when finished processing descriptor.
+ * -1 on a malformed report.
+ */
+static int get_next_hid_usage(const uint8_t *report_descriptor, uint32_t size, struct hid_usage_iterator *ctx, unsigned short *usage_page, unsigned short *usage)
+{
+	int data_len, key_size;
+	int initial = ctx->pos == 0; /* Used to handle case where no top-level application collection is defined */
+
+	int usage_found = 0;
+
+	while (ctx->pos < size) {
+		int key = report_descriptor[ctx->pos];
+		int key_cmd = key & 0xfc;
+
+		/* Determine data_len and key_size */
+		if (!get_hid_item_size(report_descriptor, size, ctx->pos, &data_len, &key_size))
+			return -1; /* malformed report */
+
+		switch (key_cmd) {
+		case 0x4: /* Usage Page 6.2.2.7 (Global) */
+			ctx->usage_page = get_hid_report_bytes(report_descriptor, size, data_len, ctx->pos);
+			ctx->usage_page_found = 1;
+			break;
+
+		case 0x8: /* Usage 6.2.2.8 (Local) */
+			if (data_len == 4) { /* Usages 5.5 / Usage Page 6.2.2.7 */
+				ctx->usage_page = get_hid_report_bytes(report_descriptor, size, 2, ctx->pos + 2);
+				ctx->usage_page_found = 1;
+				*usage = get_hid_report_bytes(report_descriptor, size, 2, ctx->pos);
+				usage_found = 1;
+			}
+			else {
+				*usage = get_hid_report_bytes(report_descriptor, size, data_len, ctx->pos);
+				usage_found = 1;
+			}
+			break;
+
+		case 0xa0: /* Collection 6.2.2.4 (Main) */
+			if (!hid_iterate_over_collection(report_descriptor, size, &ctx->pos, &data_len, &key_size)) {
+				return -1;
+			}
+
+			/* A pair is valid - to be reported when Collection is found */
+			if (usage_found && ctx->usage_page_found) {
+				*usage_page = ctx->usage_page;
+				return 0;
+			}
+
+			break;
+		}
+
+		/* Skip over this key and its associated data */
+		ctx->pos += data_len + key_size;
+	}
+
+	/* If no top-level application collection is found and usage page/usage pair is found, pair is valid
+	   https://docs.microsoft.com/en-us/windows-hardware/drivers/hid/top-level-collections */
+	if (initial && usage_found && ctx->usage_page_found) {
+		*usage_page = ctx->usage_page;
+		return 0; /* success */
+	}
+
+	return 1; /* finished processing */
+}
+
+static libusb_device_handle *hid_find_device_handle_by_bus_and_port(int bus, int addr) {
+	libusb_device **list;
+	libusb_device_handle *handler = NULL;
+	int num_devs;
+
+	num_devs = libusb_get_device_list(global_usb_context, &list);
+	if (num_devs < 0) {
+		register_global_error_format("hid_find_device_handle_by_bus_and_port: %s", strerror(errno));
```

Comment:
```suggestion
		register_global_error_format("hid_find_device_handle_by_bus_and_port: %s", libusb_strerror(num_devs));
```
libusb doesn't specify that it would set the `errno`